lib/engine/game/g_1837/game.rb

```diff
@@ -512,14 +512,20 @@ def merge_minor!(minor, corporation, allow_president_change: true)
           end
 
           if minor.cash.positive?
-            @log << "#{corporation.name} receives #{format_currency(minor.cash)}"
+            @log << "#{corporation.name} receives #{format_currency(minor.cash)} from #{minor.name}'s treasury"
             minor.spend(minor.cash, corporation)
           end
 
           unless minor.trains.empty?
-            @log << "#{corporation.name} receives #{minor.trains.map(&:name).join(', ')} train#{minor.trains.size > 1 ? 's' : ''}"
-            @round.merged_trains[corporation].concat(minor.trains)
-            minor.trains.dup.each { |t| buy_train(corporation, t, :free) }
+            trains_str = "#{minor.trains.map(&:name).join(', ')} train#{minor.trains.size > 1 ? 's' : ''}"
+            if @round.merged_trains[corporation].empty? && corporation.trains.size >= train_limit(corporation)
```

Comment:
This is highly unlikely, but it would be possible for Ug to form when each of UG1, UG2 and UG3 have two trains. Ug is going to receive four trains from the first two minors, and then the two trains from the last minor will be discarded as its train limit is four. It should receive all six trains and the president gets to decide which to keep.

Reply:
And I think you could get the same problem at the start of phase 5 if two coal companies with trains fold into the same major company.

Reply:
@ollybh when UG1 merges, won't the else case be hit and add UG1's trains to merged_trains, which would then for UG2 and UG3 have the if case evaluate to false? All 6 trains should end up in merged_trains and the player will get to pick which to choose.